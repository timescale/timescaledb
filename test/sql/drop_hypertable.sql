-- This file and its contents are licensed under the Apache License 2.0.
-- Please see the included NOTICE for copyright information and
-- LICENSE-APACHE for a copy of the license.

SELECT * from _timescaledb_catalog.hypertable;
SELECT * from _timescaledb_catalog.dimension;

CREATE TABLE should_drop (time timestamp, temp float8);
SELECT create_hypertable('should_drop', 'time');

CREATE TABLE hyper_with_dependencies (time timestamp, temp float8);
SELECT create_hypertable('hyper_with_dependencies', 'time');

CREATE VIEW dependent_view AS SELECT * FROM hyper_with_dependencies;

INSERT INTO hyper_with_dependencies VALUES (now(), 1.0);

\set ON_ERROR_STOP 0
DROP TABLE hyper_with_dependencies;
\set ON_ERROR_STOP 1
DROP TABLE hyper_with_dependencies CASCADE;
-- check that the view is dropped
SELECT oid FROM pg_class WHERE relname = 'dependent_view';

CREATE TABLE chunk_with_dependencies (time timestamp, temp float8);
SELECT create_hypertable('chunk_with_dependencies', 'time');

INSERT INTO chunk_with_dependencies VALUES (now(), 1.0);

CREATE VIEW dependent_view_chunk AS SELECT * FROM _timescaledb_internal._hyper_3_2_chunk;

\set ON_ERROR_STOP 0
DROP TABLE chunk_with_dependencies;
\set ON_ERROR_STOP 1
DROP TABLE chunk_with_dependencies CASCADE;
-- check that the view is dropped
SELECT oid FROM pg_class WHERE relname = 'dependent_view_chunk';

-- Calling create hypertable again will increment hypertable ID
-- although no new hypertable is created. Make sure we can handle this.
SELECT create_hypertable('should_drop', 'time', if_not_exists => true);
SELECT * from _timescaledb_catalog.hypertable;
SELECT * from _timescaledb_catalog.dimension;
DROP TABLE should_drop;

CREATE TABLE should_drop (time timestamp, temp float8);
SELECT create_hypertable('should_drop', 'time');

INSERT INTO should_drop VALUES (now(), 1.0);
SELECT * from _timescaledb_catalog.hypertable;
SELECT * from _timescaledb_catalog.dimension;

