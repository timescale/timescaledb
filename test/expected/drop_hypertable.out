-- This file and its contents are licensed under the Apache License 2.0.
-- Please see the included NOTICE for copyright information and
-- LICENSE-APACHE for a copy of the license.
SELECT * from _timescaledb_catalog.hypertable;
 id | schema_name | table_name | associated_schema_name | associated_table_prefix | num_dimensions | chunk_sizing_func_schema | chunk_sizing_func_name | chunk_target_size | compression_state | compressed_hypertable_id | status 
----+-------------+------------+------------------------+-------------------------+----------------+--------------------------+------------------------+-------------------+-------------------+--------------------------+--------
(0 rows)

SELECT * from _timescaledb_catalog.dimension;
 id | hypertable_id | column_name | column_type | aligned | num_slices | partitioning_func_schema | partitioning_func | interval_length | compress_interval_length | integer_now_func_schema | integer_now_func | type 
----+---------------+-------------+-------------+---------+------------+--------------------------+-------------------+-----------------+--------------------------+-------------------------+------------------+------
(0 rows)

CREATE TABLE should_drop (time timestamp, temp float8);
SELECT create_hypertable('should_drop', 'time');
WARNING:  column type "timestamp without time zone" used for "time" does not follow best practices
NOTICE:  adding not-null constraint to column "time"
    create_hypertable     
--------------------------
 (1,public,should_drop,t)
(1 row)

CREATE TABLE hyper_with_dependencies (time timestamp, temp float8);
SELECT create_hypertable('hyper_with_dependencies', 'time');
WARNING:  column type "timestamp without time zone" used for "time" does not follow best practices
NOTICE:  adding not-null constraint to column "time"
          create_hypertable           
--------------------------------------
 (2,public,hyper_with_dependencies,t)
(1 row)

CREATE VIEW dependent_view AS SELECT * FROM hyper_with_dependencies;
INSERT INTO hyper_with_dependencies VALUES (now(), 1.0);
\set ON_ERROR_STOP 0
DROP TABLE hyper_with_dependencies;
ERROR:  cannot drop table hyper_with_dependencies because other objects depend on it
\set ON_ERROR_STOP 1
DROP TABLE hyper_with_dependencies CASCADE;
NOTICE:  drop cascades to view dependent_view
\dv
      List of relations
 Schema | Name | Type | Owner 
--------+------+------+-------
(0 rows)

CREATE TABLE chunk_with_dependencies (time timestamp, temp float8);
SELECT create_hypertable('chunk_with_dependencies', 'time');
WARNING:  column type "timestamp without time zone" used for "time" does not follow best practices
NOTICE:  adding not-null constraint to column "time"
          create_hypertable           
--------------------------------------
 (3,public,chunk_with_dependencies,t)
(1 row)

INSERT INTO chunk_with_dependencies VALUES (now(), 1.0);
CREATE VIEW dependent_view_chunk AS SELECT * FROM _timescaledb_internal._hyper_3_2_chunk;
\set ON_ERROR_STOP 0
DROP TABLE chunk_with_dependencies;
ERROR:  cannot drop table _timescaledb_internal._hyper_3_2_chunk because other objects depend on it
\set ON_ERROR_STOP 1
DROP TABLE chunk_with_dependencies CASCADE;
NOTICE:  drop cascades to view dependent_view_chunk
\dv
      List of relations
 Schema | Name | Type | Owner 
--------+------+------+-------
(0 rows)

-- Calling create hypertable again will increment hypertable ID
-- although no new hypertable is created. Make sure we can handle this.
SELECT create_hypertable('should_drop', 'time', if_not_exists => true);
NOTICE:  table "should_drop" is already a hypertable, skipping
    create_hypertable     
--------------------------
 (1,public,should_drop,f)
(1 row)

SELECT * from _timescaledb_catalog.hypertable;
 id | schema_name | table_name  | associated_schema_name | associated_table_prefix | num_dimensions | chunk_sizing_func_schema |  chunk_sizing_func_name  | chunk_target_size | compression_state | compressed_hypertable_id | status 
----+-------------+-------------+------------------------+-------------------------+----------------+--------------------------+--------------------------+-------------------+-------------------+--------------------------+--------
  1 | public      | should_drop | _timescaledb_internal  | _hyper_1                |              1 | _timescaledb_functions   | calculate_chunk_interval |                 0 |                 0 |                          |      0
(1 row)

SELECT * from _timescaledb_catalog.dimension;
 id | hypertable_id | column_name |         column_type         | aligned | num_slices | partitioning_func_schema | partitioning_func | interval_length | compress_interval_length | integer_now_func_schema | integer_now_func | type 
----+---------------+-------------+-----------------------------+---------+------------+--------------------------+-------------------+-----------------+--------------------------+-------------------------+------------------+------
  1 |             1 | time        | timestamp without time zone | t       |            |                          |                   |    604800000000 |                          |                         |                  | o
(1 row)

DROP TABLE should_drop;
CREATE TABLE should_drop (time timestamp, temp float8);
SELECT create_hypertable('should_drop', 'time');
WARNING:  column type "timestamp without time zone" used for "time" does not follow best practices
NOTICE:  adding not-null constraint to column "time"
    create_hypertable     
--------------------------
 (4,public,should_drop,t)
(1 row)

INSERT INTO should_drop VALUES (now(), 1.0);
SELECT * from _timescaledb_catalog.hypertable;
 id | schema_name | table_name  | associated_schema_name | associated_table_prefix | num_dimensions | chunk_sizing_func_schema |  chunk_sizing_func_name  | chunk_target_size | compression_state | compressed_hypertable_id | status 
----+-------------+-------------+------------------------+-------------------------+----------------+--------------------------+--------------------------+-------------------+-------------------+--------------------------+--------
  4 | public      | should_drop | _timescaledb_internal  | _hyper_4                |              1 | _timescaledb_functions   | calculate_chunk_interval |                 0 |                 0 |                          |      0
(1 row)

SELECT * from _timescaledb_catalog.dimension;
 id | hypertable_id | column_name |         column_type         | aligned | num_slices | partitioning_func_schema | partitioning_func | interval_length | compress_interval_length | integer_now_func_schema | integer_now_func | type 
----+---------------+-------------+-----------------------------+---------+------------+--------------------------+-------------------+-----------------+--------------------------+-------------------------+------------------+------
  4 |             4 | time        | timestamp without time zone | t       |            |                          |                   |    604800000000 |                          |                         |                  | o
(1 row)

